### ISSUE TITLE
incorrect method sorting with union in typevar bound

### ISSUE THREAD

**Issue by srp** [JuliaLang/julia#8915]
The documentation states that methods are selected based off which is the most specific. While not explicitly stated, this generally also seems to be true of constructors. However on 0.4, the following seems to be functioning off the order the constructors are declared in, rather then which is most specific:

``` Julia
type A{O <: Union(Float64,String)}
    A(a) = println("A(a)")
    A(a::Int) = println("A(a::Int)")
end

type B{O <: Union(Float64,String)}
    B(a::Int) = println("B(a::Int)")
    B(a) = println("B(a)")
end

A{Float64}(1)
B{Float64}(1)
```

Which produces the following output:

```
A(a)
B(a::Int)
```

I can't reproduce this on 0.3, only 0.4. I've also only been able to reproduce this when parameterizing  the type with a Union.

**Comment by timholy** [comment]
Another likely instance of this can be seen in ColorTypes, depending on whether we precompile or not:
```julia
tim@diva:~/.julia/v0.5/ColorTypes/test$ julia -q
julia> using ColorTypes

julia> Gray(0.8)
Gray{Float64}(0.8)

julia> 
tim@diva:~/.julia/v0.5/ColorTypes/test$ julia -q
julia> include("../src/ColorTypes.jl")
ColorTypes

julia> using ColorTypes

julia> Gray(0.8)
------------------------------------------------------------------------------------------
StackOverflowError                                      Stacktrace (most recent call last)
[#1] — ColorTypes.Gray{T<:Union{AbstractFloat,Bool,FixedPointNumbers.FixedPoint}}(::Float64)
       ⌙ at types.jl:369 (repeats 80000 times)

StackOverflowError:
```

**Comment by JeffBezanson** [comment]
closed by #18457
